In [1]:
import numpy as np
import pandas as pd 

In [2]:
# pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [3]:
df = pd.read_parquet("../../../Data/LLCP2022.parquet")

In [4]:
df

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
1          1.0     1.0  b'02042022'  b'02'  b'04'  b'2022'    1100.0   
2          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
3          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
4          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
...        ...     ...          ...    ...    ...      ...       ...   
445127    78.0    11.0  b'12192022'  b'12'  b'19'  b'2022'    1100.0   
445128    78.0    11.0  b'12212022'  b'12'  b'21'  b'2022'    1100.0   
445129    78.0    11.0  b'11292022'  b'11'  b'29'  b'2022'    1100.0   
445130    78.0    11.0  b'12082022'  b'12'  b'08'  b'2022'    1100.0   
445131    78.0    11.0  b'12142022'  b'12'  b'14'  b'2022'    1100.0   

                SEQNO          _PSU  CTELENM1  PVTRESD1  COLGHOUS  STATERE1  \
0       b'2022000001'  2.022000e+09       1.0       1.0       NaN       1.0   
1       b'2022000002'  2.022000e+09       1.0       1.0       NaN       1.0   
2       b'2022000003'  2.022000e+09       1.0       1.0       NaN       1.0   
3       b'2022000004'  2.022000e+09       1.0       1.0       NaN       1.0   
4       b'2022000005'  2.022000e+09       1.0       1.0       NaN       1.0   
...               ...           ...       ...       ...       ...       ...   
445127  b'2022001527'  2.022002e+09       NaN       NaN       NaN       NaN   
445128  b'2022001528'  2.022002e+09       NaN       NaN       NaN       NaN   
445129  b'2022001529'  2.022002e+09       NaN       NaN       NaN       NaN   
445130  b'2022001530'  2.022002e+09       NaN       NaN       NaN       NaN   
445131  b'2022001531'  2.022002e+09       NaN       NaN       NaN       NaN   

        CELPHON1  LADULT1  COLGSEX1  NUMADULT  LANDSEX1  NUMMEN  NUMWOMEN  \
0            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
1            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
2            2.0      1.0       NaN       1.0       2.0     NaN       NaN   
3            2.0      1.0       NaN       3.0       NaN     2.0       1.0   
4            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
...          ...      ...       ...       ...       ...     ...       ...   
445127       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445128       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445129       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445130       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445131       NaN      NaN       NaN       NaN       NaN     NaN       NaN   

        RESPSLCT  SAFETIME  CTELNUM1  CELLFON5  CADULT1  CELLSEX1  PVTRESD3  \
0            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
1            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
2            NaN       NaN       NaN       NaN      NaN       NaN       NaN   
3            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
4            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
...          ...       ...       ...       ...      ...       ...       ...   
445127       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445128       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445129       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445130       NaN       1.0       1.0       1.0      1.0       1.0       1.0   
445131       NaN       1.0       1.0       1.0      1.0       1.0       1.0   

        CCLGHOUS  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  PHYSHLTH  \
0            NaN      NaN       NaN      NaN     2.0      2.0      88.0   
1            NaN      NaN       NaN      NaN     2.0      1.0      88.0   
2            NaN      NaN       NaN      NaN     2.0      2.0       2.0   
3            NaN    

In [5]:
df.replace(to_replace='None', value=np.nan, inplace=True)
df = df.dropna(axis=1, how='all')

In [6]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
columns_with_missing_values = missing_percentage[missing_percentage > 50]

result_df = pd.DataFrame({
    'Column': columns_with_missing_values.index,
    'Percentage_of_Missing_Values': columns_with_missing_values.values.round(2)
})
result_df.sort_values(by= 'Percentage_of_Missing_Values', ascending=False)

Column  Percentage_of_Missing_Values
2    COLGHOUS                        100.00
6    COLGSEX1                        100.00
12   CCLGHOUS                         99.70
85   CSRVCTL2                         99.65
60   HPVADSHT                         99.62
..        ...                           ...
18     HADMAM                         50.17
169  _CLNSCP1                         50.14
176  _CRCREC2                         50.11
170  _SGMSCP1                         50.08
171  _SGMS101                         50.08

[184 rows x 2 columns]

In [7]:
df = df.drop(columns = columns_with_missing_values.index, axis=1)
df

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
1          1.0     1.0  b'02042022'  b'02'  b'04'  b'2022'    1100.0   
2          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
3          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
4          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
...        ...     ...          ...    ...    ...      ...       ...   
445127    78.0    11.0  b'12192022'  b'12'  b'19'  b'2022'    1100.0   
445128    78.0    11.0  b'12212022'  b'12'  b'21'  b'2022'    1100.0   
445129    78.0    11.0  b'11292022'  b'11'  b'29'  b'2022'    1100.0   
445130    78.0    11.0  b'12082022'  b'12'  b'08'  b'2022'    1100.0   
445131    78.0    11.0  b'12142022'  b'12'  b'14'  b'2022'    1100.0   

                SEQNO          _PSU  SAFETIME  CTELNUM1  CELLFON5  CADULT1  \
0       b'2022000001'  2.022000e+09       NaN       NaN       NaN      NaN   
1       b'2022000002'  2.022000e+09       NaN       NaN       NaN      NaN   
2       b'2022000003'  2.022000e+09       NaN       NaN       NaN      NaN   
3       b'2022000004'  2.022000e+09       NaN       NaN       NaN      NaN   
4       b'2022000005'  2.022000e+09       NaN       NaN       NaN      NaN   
...               ...           ...       ...       ...       ...      ...   
445127  b'2022001527'  2.022002e+09       1.0       1.0       1.0      1.0   
445128  b'2022001528'  2.022002e+09       1.0       1.0       1.0      1.0   
445129  b'2022001529'  2.022002e+09       1.0       1.0       1.0      1.0   
445130  b'2022001530'  2.022002e+09       1.0       1.0       1.0      1.0   
445131  b'2022001531'  2.022002e+09       1.0       1.0       1.0      1.0   

        CELLSEX1  PVTRESD3  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  \
0            NaN       NaN      NaN       NaN      NaN     2.0      2.0   
1            NaN       NaN      NaN       NaN      NaN     2.0      1.0   
2            NaN       NaN      NaN       NaN      NaN     2.0      2.0   
3            NaN       NaN      NaN       NaN      NaN     2.0      1.0   
4            NaN       NaN      NaN       NaN      NaN     2.0      4.0   
...          ...       ...      ...       ...      ...     ...      ...   
445127       2.0       1.0      1.0       2.0      5.0     2.0      3.0   
445128       2.0       1.0      1.0       1.0      2.0     2.0      1.0   
445129       2.0       1.0      1.0       2.0      1.0     2.0      5.0   
445130       1.0       1.0      1.0       2.0      2.0     1.0      2.0   
445131       1.0       1.0      1.0       2.0      1.0     1.0      2.0   

        PHYSHLTH  MENTHLTH  POORHLTH  PRIMINSR  PERSDOC3  MEDCOST1  CHECKUP1  \
0           88.0      88.0       NaN      99.0       1.0       2.0       1.0   
1           88.0      88.0       NaN       3.0       2.0       2.0       8.0   
2            2.0       3.0       2.0       1.0       1.0       2.0       1.0   
3           88.0      88.0       NaN      99.0       1.0       2.0       1.0   
4            2.0      88.0      88.0       7.0       2.0       2.0       1.0   
...          ...       ...       ...       ...       ...       ...       ...   
445127      88.0       3.0      30.0       3.0       3.0       2.0       2.0   
445128       2.0       2.0      88.0       1.0       2.0       2.0       1.0   
445129      30.0      30.0      30.0      88.0       3.0       1.0       4.0   
445130      88.0      88.0       NaN       3.0       2.0       2.0       1.0   
445131      88.0       1.0      88.0      88.0       3.0       1.0       8.0   

        EXERANY2  SLEPTIM1  LASTDEN4  RMVTETH4  CVDINFR4  CVDCRHD4  CVDSTRK3  \
0            2.0       8.0       NaN       NaN       2.0       2.0       2.0   
1            2.0       6.0       NaN       NaN       2.0       2.0       2.0   
2            1.0       5.0       NaN       NaN       2.0       2.0       2.0   
3            1.0       7

In [8]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
columns_with_missing_values = missing_percentage[missing_percentage >= 0]

result_df = pd.DataFrame({
    'Column': columns_with_missing_values.index,
    'Percentage_of_Missing_Values': columns_with_missing_values.values.round(2)
})
result_df.sort_values(by= 'Percentage_of_Missing_Values', ascending=False)

Column  Percentage_of_Missing_Values
80  SDHSTRE1                         43.56
79  SDHTRNSP                         43.40
78  SDHUTILS                         43.32
77  SDHBILLS                         43.27
76  SDHFOOD1                         43.20
..       ...                           ...
7      SEQNO                          0.00
8       _PSU                          0.00
18    SEXVAR                          0.00
19   GENHLTH                          0.00
0     _STATE                          0.00

[141 rows x 2 columns]

If you keep <= 40 then you get 122820 rows × 129 columns.    
If you keep columns with <= 50 you get 33295 rows × 141 columns.    
So I am going with 40.    

In [9]:
df = df[result_df[result_df['Percentage_of_Missing_Values'] <= 40.0].Column].dropna()
invalid_columns = ['CVDCRHD4', 'CVDINFR4', 'LCSCTSC1']
# These columns are being removed because they are asking the same information as the target and should not skew the weights
df = df.drop(invalid_columns ,axis =1)
df

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
2087       1.0     5.0  b'05202022'  b'05'  b'20'  b'2022'    1100.0   
2088       1.0     5.0  b'05102022'  b'05'  b'10'  b'2022'    1100.0   
2090       1.0     5.0  b'06072022'  b'06'  b'07'  b'2022'    1100.0   
2091       1.0     5.0  b'05262022'  b'05'  b'26'  b'2022'    1100.0   
2093       1.0     5.0  b'05102022'  b'05'  b'10'  b'2022'    1100.0   
...        ...     ...          ...    ...    ...      ...       ...   
431675    55.0    12.0  b'01122023'  b'01'  b'12'  b'2023'    1100.0   
431676    55.0    12.0  b'01172023'  b'01'  b'17'  b'2023'    1100.0   
431677    55.0    12.0  b'01092023'  b'01'  b'09'  b'2023'    1100.0   
431680    55.0     7.0  b'07222022'  b'07'  b'22'  b'2022'    1200.0   
431682    55.0     8.0  b'08302022'  b'08'  b'30'  b'2022'    1200.0   

                SEQNO          _PSU  SAFETIME  CTELNUM1  CELLFON5  CADULT1  \
2087    b'2022002088'  2.022002e+09       1.0       1.0       1.0      1.0   
2088    b'2022002089'  2.022002e+09       1.0       1.0       1.0      1.0   
2090    b'2022002091'  2.022002e+09       1.0       1.0       1.0      1.0   
2091    b'2022002092'  2.022002e+09       1.0       1.0       1.0      1.0   
2093    b'2022002094'  2.022002e+09       1.0       1.0       1.0      1.0   
...               ...           ...       ...       ...       ...      ...   
431675  b'2022011268'  2.022011e+09       1.0       1.0       1.0      1.0   
431676  b'2022011269'  2.022011e+09       1.0       1.0       1.0      1.0   
431677  b'2022011270'  2.022011e+09       1.0       1.0       1.0      1.0   
431680  b'2022011273'  2.022011e+09       1.0       1.0       1.0      1.0   
431682  b'2022011275'  2.022011e+09       1.0       1.0       1.0      1.0   

        CELLSEX1  PVTRESD3  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  \
2087         2.0       1.0      1.0       2.0      1.0     2.0      4.0   
2088         1.0       1.0      1.0       1.0      2.0     1.0      4.0   
2090         2.0       1.0      1.0       2.0      2.0     2.0      2.0   
2091         2.0       1.0      1.0       2.0      2.0     2.0      4.0   
2093         1.0       1.0      1.0       2.0      4.0     1.0      3.0   
...          ...       ...      ...       ...      ...     ...      ...   
431675       2.0       1.0      1.0       2.0      1.0     2.0      2.0   
431676       2.0       1.0      1.0       1.0      2.0     2.0      2.0   
431677       1.0       1.0      1.0       2.0      3.0     1.0      2.0   
431680       2.0       1.0      2.0       2.0      2.0     2.0      1.0   
431682       1.0       1.0      2.0       2.0      2.0     1.0      2.0   

        PHYSHLTH  MENTHLTH  PRIMINSR  PERSDOC3  MEDCOST1  CHECKUP1  EXERANY2  \
2087         3.0      10.0       1.0       2.0       2.0       2.0       1.0   
2088        30.0      30.0       3.0       2.0       2.0       1.0       2.0   
2090        88.0      88.0       1.0       1.0       2.0       8.0       1.0   
2091         2.0      88.0       1.0       1.0       2.0       1.0       1.0   
2093         1.0      88.0      88.0       3.0       1.0       4.0       1.0   
...          ...       ...       ...       ...       ...       ...       ...   
431675       2.0      88.0       3.0       1.0       2.0       1.0       2.0   
431676      88.0      88.0       1.0       2.0       2.0       1.0       1.0   
431677       5.0      88.0       1.0       1.0       2.0       1.0       1.0   
431680       1.0      88.0       1.0       2.0       2.0       1.0       1.0   
431682      88.0      88.0       3.0       2.0       2.0       1.0       1.0   

        SLEPTIM1  LASTDEN4  RMVTETH4  CVDSTRK3  ASTHMA3  CHCSCNC1  CHCOCNC1  \
2087         8.0       2.0       8.0       2.0      2.0       2.0       2.0   
2088         6.0       2.0       3.0       2.0      1.0       1.0       1.0   
2090         6.0       2.0       1.0       2.0      1.0       2.0       2.0   
2091         7.0       2.0  

In [10]:
target = df._MICHD
df = df.drop('_MICHD', axis=1)
target

2087      2.0
2088      1.0
2090      2.0
2091      2.0
2093      2.0
         ... 
431675    2.0
431676    2.0
431677    2.0
431680    2.0
431682    2.0
Name: _MICHD, Length: 122820, dtype: float64

In [11]:
# value_counts_per_column = df.apply(lambda x: x.value_counts())

# # Step 2: Check for columns where the unique values are 1.0 and 2.0 only
# columns_with_1_and_2 = value_counts_per_column.columns[(value_counts_per_column.loc[1.0] + value_counts_per_column.loc[2.0]).eq(value_counts_per_column.sum(axis=0))]

# # Step 3: Display the columns
# print("Columns with value counts of 1.0 and 2.0 only:")
# print(columns_with_1_and_2)

In [12]:
# value_counts_per_column = df.apply(lambda x: x.value_counts())
# columns_with_4_or_less_unique = value_counts_per_column.columns[value_counts_per_column.nunique() <= 4]

# print(columns_with_4_or_less_unique)

In [13]:
# for col in columns_with_4_or_less_unique:
#     print(df[col].value_counts())

In [14]:
unique_counts_df = pd.DataFrame({
    'Column Name': df.columns,
    'Number of Unique Values': df.nunique()
})
unique_counts_df.sort_values(by='Number of Unique Values', ascending=False)

Column Name  Number of Unique Values
_LLCPWT      _LLCPWT                    68726
SEQNO          SEQNO                    21541
_PSU            _PSU                    21541
_LLCPWT2    _LLCPWT2                     3228
_BMI5          _BMI5                     2902
...              ...                      ...
CADULT1      CADULT1                        1
CELLFON5    CELLFON5                        1
CTELNUM1    CTELNUM1                        1
SAFETIME    SAFETIME                        1
_RAWRAKE    _RAWRAKE                        1

[125 rows x 2 columns]

In [15]:
# colsW10OrLessGroups = unique_counts_df[unique_counts_df['Number of Unique Values'] <10]['Column Name']
# for col in colsW10OrLessGroups:
#     print(df[col].value_counts())

# print("Number of Columns with Less than 10 unique values:  ", len(colsW10OrLessGroups))

In [16]:
# colsW10To20Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 10) & (unique_counts_df['Number of Unique Values'] < 20)]['Column Name']

# for col in colsW10To20Groups:
#     print(df[col].value_counts())

# print("Number of columns with 10 - 20 unique values:   ", len(colsW10To20Groups))

So CHILDREN and _STATE are the two categorical values and everything over 20 unique values is a numerical value. 

In [17]:
# colsW20To30Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 20) & (unique_counts_df['Number of Unique Values'] < 30)]['Column Name']

# for col in colsW20To30Groups:
#     print(df[col].value_counts())
# print("Number of columns with 20 - 30 unique values:   ", len(colsW20To30Groups))

In [18]:
# colsW30To40Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 30) & (unique_counts_df['Number of Unique Values'] < 40)]['Column Name']

# for col in colsW30To40Groups:
#     print(df[col].value_counts())
# print("Number of columns with 30 - 40 unique values:   ", len(colsW30To40Groups))

In [19]:
# colsW40To50Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 40) & (unique_counts_df['Number of Unique Values'] < 50)]['Column Name']

# for col in colsW40To50Groups:
#     print(df[col].value_counts())
# print("Number of columns with 40 - 50 unique values:   ", len(colsW40To50Groups))

In [20]:
# colsW50To60Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 50) & (unique_counts_df['Number of Unique Values'] < 60)]['Column Name']

# for col in colsW50To60Groups:
#     print(df[col].value_counts())
# print("Number of columns with 50 - 60 unique values:   ", len(colsW50To60Groups))

In [21]:
# colsW60To70Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 60) & (unique_counts_df['Number of Unique Values'] < 70)]['Column Name']

# for col in colsW60To70Groups:
#     print(df[col].value_counts())
# print("Number of columns with 60 - 70 unique values:   ", len(colsW60To70Groups))

In [22]:
# colsW70To1000Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 70) & (unique_counts_df['Number of Unique Values'] < 1000)]['Column Name']

# for col in colsW70To1000Groups:
#     print(df[col].value_counts())
# print("Number of columns with 70 - 1000 unique values:   ", len(colsW70To1000Groups))

In [23]:
# colsW1000To80Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 1000) & (unique_counts_df['Number of Unique Values'] < 60000)]['Column Name']

# for col in colsW1000To80Groups:
#     print(df[col].value_counts())
# print("Number of columns with 1000 - 80 unique values:   ", len(colsW1000To80Groups))

In [24]:
unique_counts_df[(unique_counts_df['Number of Unique Values'] <=20)]['Column Name']

FMONTH        FMONTH
IMONTH        IMONTH
IYEAR          IYEAR
DISPCODE    DISPCODE
SAFETIME    SAFETIME
              ...   
_SMOKGRP    _SMOKGRP
DRNKANY6    DRNKANY6
_RFBING6    _RFBING6
_RFDRHV8    _RFDRHV8
_AIDTST4    _AIDTST4
Name: Column Name, Length: 101, dtype: object

In [25]:
categorical_columns =["_STATE"]
filtered_columns = list(unique_counts_df[(unique_counts_df['Number of Unique Values'] <= 20)]['Column Name'])

# # Ensure that filtered_columns are present in df
# filtered_columns = [col for col in filtered_columns if col in df.columns]

df_categorical = df[categorical_columns + filtered_columns]
# df_categorical = df_categorical.drop('_MICHD', axis=1)


In [26]:
df_categorical

_STATE  FMONTH IMONTH    IYEAR  DISPCODE  SAFETIME  CTELNUM1  \
2087       1.0     5.0  b'05'  b'2022'    1100.0       1.0       1.0   
2088       1.0     5.0  b'05'  b'2022'    1100.0       1.0       1.0   
2090       1.0     5.0  b'06'  b'2022'    1100.0       1.0       1.0   
2091       1.0     5.0  b'05'  b'2022'    1100.0       1.0       1.0   
2093       1.0     5.0  b'05'  b'2022'    1100.0       1.0       1.0   
...        ...     ...    ...      ...       ...       ...       ...   
431675    55.0    12.0  b'01'  b'2023'    1100.0       1.0       1.0   
431676    55.0    12.0  b'01'  b'2023'    1100.0       1.0       1.0   
431677    55.0    12.0  b'01'  b'2023'    1100.0       1.0       1.0   
431680    55.0     7.0  b'07'  b'2022'    1200.0       1.0       1.0   
431682    55.0     8.0  b'08'  b'2022'    1200.0       1.0       1.0   

        CELLFON5  CADULT1  CELLSEX1  PVTRESD3  CSTATE1  LANDLINE  SEXVAR  \
2087         1.0      1.0       2.0       1.0      1.0       2.0     2.0   
2088         1.0      1.0       1.0       1.0      1.0       1.0     1.0   
2090         1.0      1.0       2.0       1.0      1.0       2.0     2.0   
2091         1.0      1.0       2.0       1.0      1.0       2.0     2.0   
2093         1.0      1.0       1.0       1.0      1.0       2.0     1.0   
...          ...      ...       ...       ...      ...       ...     ...   
431675       1.0      1.0       2.0       1.0      1.0       2.0     2.0   
431676       1.0      1.0       2.0       1.0      1.0       1.0     2.0   
431677       1.0      1.0       1.0       1.0      1.0       2.0     1.0   
431680       1.0      1.0       2.0       1.0      2.0       2.0     2.0   
431682       1.0      1.0       1.0       1.0      2.0       2.0     1.0   

        GENHLTH  PRIMINSR  PERSDOC3  MEDCOST1  CHECKUP1  EXERANY2  LASTDEN4  \
2087        4.0       1.0       2.0       2.0       2.0       1.0       2.0   
2088        4.0       3.0       2.0       2.0       1.0       2.0       2.0   
2090        2.0       1.0       1.0       2.0       8.0       1.0       2.0   
2091        4.0       1.0       1.0       2.0       1.0       1.0       2.0   
2093        3.0      88.0       3.0       1.0       4.0       1.0       8.0   
...         ...       ...       ...       ...       ...       ...       ...   
431675      2.0       3.0       1.0       2.0       1.0       2.0       1.0   
431676      2.0       1.0       2.0       2.0       1.0       1.0       1.0   
431677      2.0       1.0       1.0       2.0       1.0       1.0       1.0   
431680      1.0       1.0       2.0       2.0       1.0       1.0       1.0   
431682      2.0       3.0       2.0       2.0       1.0       1.0       1.0   

        RMVTETH4  CVDSTRK3  ASTHMA3  CHCSCNC1  CHCOCNC1  CHCCOPD3  ADDEPEV3  \
2087         8.0       2.0      2.0       2.0       2.0       1.0       2.0   
2088         3.0       2.0      1.0       1.0       1.0       1.0       1.0   
2090         1.0       2.0      1.0       2.0       2.0       2.0       2.0   
2091         8.0       2.0      2.0       2.0       2.0       2.0       2.0   
2093         2.0       2.0      2.0       2.0       2.0       1.0       2.0   
...          ...       ...      ...       ...       ...       ...       ...   
431675       1.0       2.0      2.0       2.0       2.0       2.0       2.0   
431676       8.0       2.0      2.0       1.0       1.0       2.0       2.0   
431677       8.0       2.0      2.0       2.0       1.0       2.0       2.0   
431680       8.0       2.0      1.0       2.0       2.0       2.0       1.0   
431682       8.0       2.0      2.0       2.0       2.0       2.0       2.0   

        CHCKDNY2  HAVARTH4  DIABETE4  MARITAL  EDUCA  RENTHOM1  CPDEMO1C  \
2087         2.0       2.0       3.0      5.0    6.0       2.0       2.0   
2088         2.0       1.0       1.0      1.0    3.0       2.0       1.0   
2090         2.0       1.0       3.0      1.0    4.0       1.0       1.0   
2091         2.0       1.0  

In [27]:
cols = list(df_categorical.columns)
mask = unique_counts_df['Column Name'].isin(cols)
filtered_rows = unique_counts_df[mask]
filtered_rows.sort_values(by = ['Number of Unique Values'], ascending = False)

Column Name  Number of Unique Values
_STATE        _STATE                       32
CHILDREN    CHILDREN                       18
INCOME3      INCOME3                       13
PRIMINSR    PRIMINSR                       13
IMONTH        IMONTH                       12
...              ...                      ...
CADULT1      CADULT1                        1
CELLFON5    CELLFON5                        1
CTELNUM1    CTELNUM1                        1
SAFETIME    SAFETIME                        1
_RAWRAKE    _RAWRAKE                        1

[102 rows x 2 columns]

In [28]:
df_categorical = df_categorical.astype('category')
df_categorical.dtypes

_STATE      category
FMONTH      category
IMONTH      category
IYEAR       category
DISPCODE    category
              ...   
_SMOKGRP    category
DRNKANY6    category
_RFBING6    category
_RFDRHV8    category
_AIDTST4    category
Length: 102, dtype: object

In [29]:
numerical_columns = list(set(df.columns) - set(df_categorical.columns))
df_numerical = df[numerical_columns]
df_numerical = df_numerical.drop(['IDAY', 'IDATE', 'SEQNO'], axis=1)
df_numerical

_LLCPWT      _DRNKWK2  HHADULT  MENTHLTH   _BMI5   HTM4  \
2087     867.513629  1.170000e+02      1.0      10.0  3828.0  163.0   
2088    2071.628967  5.397605e-79      2.0      30.0  2889.0  173.0   
2090    1935.196682  5.397605e-79      2.0      88.0  3292.0  157.0   
2091    1752.329641  4.700000e+01      2.0      88.0  2850.0  170.0   
2093    1666.019237  4.900000e+03      4.0      88.0  2067.0  175.0   
...             ...           ...      ...       ...     ...    ...   
431675   130.521260  1.000000e+02      1.0      88.0  2986.0  168.0   
431676   580.239671  3.500000e+02      2.0      88.0  2712.0  168.0   
431677   629.560760  3.000000e+02      3.0      88.0  3138.0  180.0   
431680   553.207932  7.000000e+01      2.0      88.0  2378.0  157.0   
431682   182.212572  1.170000e+02      2.0      88.0  2585.0  173.0   

          _WT2RAKE    _STSTR  ALCDAY4      _STRWT      DROCDY4_  SLEPTIM1  \
2087    193.714430   12011.0    205.0  193.714430  1.700000e+01       8.0   
2088    193.714430   12011.0    888.0  193.714430  5.397605e-79       6.0   
2090    193.714430   12011.0    888.0  193.714430  5.397605e-79       6.0   
2091    193.714430   12011.0    202.0  193.714430  7.000000e+00       7.0   
2093    193.714430   12012.0    230.0  193.714430  1.000000e+02       6.0   
...            ...       ...      ...         ...           ...       ...   
431675   16.699425  552101.0    101.0   16.699425  1.400000e+01       7.0   
431676   16.699425  552102.0    215.0   16.699425  5.000000e+01       8.0   
431677   16.699425  552101.0    103.0   16.699425  4.300000e+01       8.0   
431680   27.905265  552022.0    203.0   27.905265  1.000000e+01       6.0   
431682   15.609421  552041.0    205.0   15.609421  1.700000e+01       8.0   

        HTIN4     _LLCPWT2  HEIGHT3  PHYSHLTH    WTKG3  WEIGHT2          _PSU  \
2087     64.0  1809.103461    504.0       3.0  10115.0    223.0  2.022002e+09   
2088     68.0  1941.776707    508.0      30.0   8618.0    190.0  2.022002e+09   
2090     62.0  1809.103461    502.0      88.0   8165.0    180.0  2.022002e+09   
2091     67.0  1809.103461    507.0       2.0   8255.0    182.0  2.022002e+09   
2093     69.0  1809.103461    509.0       1.0   6350.0    140.0  2.022002e+09   
...       ...          ...      ...       ...      ...      ...           ...   
431675   66.0   557.869736    506.0       2.0   8391.0    185.0  2.022011e+09   
431676   66.0   813.831680    506.0      88.0   7620.0    168.0  2.022011e+09   
431677   71.0   557.869736    511.0       5.0  10206.0    225.0  2.022011e+09   
431680   62.0   796.997215    502.0       1.0   5897.0    130.0  2.022011e+09   
431682   68.0   431.619384    508.0      88.0   7711.0    170.0  2.022011e+09   

        _AGE80  
2087      52.0  
2088      66.0  
2090      61.0  
2091      48.0  
2093      55.0  
...        ...  
431675    68.0  
431676    62.0  
431677    50.0  
431680    47.0  
431682    71.0  

[122820 rows x 20 columns]

In [30]:
df_numerical.dtypes.sort_values()

_LLCPWT     float64
WEIGHT2     float64
WTKG3       float64
PHYSHLTH    float64
HEIGHT3     float64
_LLCPWT2    float64
HTIN4       float64
SLEPTIM1    float64
DROCDY4_    float64
_STRWT      float64
ALCDAY4     float64
_STSTR      float64
_WT2RAKE    float64
HTM4        float64
_BMI5       float64
MENTHLTH    float64
HHADULT     float64
_DRNKWK2    float64
_PSU        float64
_AGE80      float64
dtype: object

In [31]:
# # Get the total number of columns
# total_columns = df.shape[1]

# # Calculate the indices to split the columns into eight equal parts
# eighth1 = total_columns // 8
# eighth2 = 2 * total_columns // 8
# eighth3 = 3 * total_columns // 8
# eighth4 = 4 * total_columns // 8
# eighth5 = 5 * total_columns // 8
# eighth6 = 6 * total_columns // 8
# eighth7 = 7 * total_columns // 8

# # Slice the DataFrame to get the eight equal parts
# part1 = df_categorical.iloc[:, :eighth1]
# part2 = df_categorical.iloc[:, eighth1:eighth2]
# part3 = df_categorical.iloc[:, eighth2:eighth3]
# part4 = df_categorical.iloc[:, eighth3:eighth4]
# part5 = df_categorical.iloc[:, eighth4:eighth5]
# part6 = df_categorical.iloc[:, eighth5:eighth6]
# part7 = df_categorical.iloc[:, eighth6:eighth7]
# parts = [part1, part2, part3, part4, part5, part6, part7]

# print(part1.shape, part2.shape, part3.shape, part4.shape, part5.shape, part6.shape, part7.shape)
# total_col = 0
# for part in parts:
#     total_col += part.shape[1]
# print(total_col, df_categorical.shape)

In [32]:
# index = 0
# for part in parts:
#     part = pd.get_dummies(part)
#     print(part.shape)
#     parts[index] = part
#     index+=1



In [33]:
# df_categorical = pd.concat(parts, axis=1)
# df_categorical.shape

In [34]:
duplicate_columns = df_categorical.columns[df_categorical.columns.duplicated()]
duplicate_columns

Index([], dtype='object')

In [35]:
df_categorical = pd.get_dummies(df_categorical)
df_categorical

_STATE_1.0  _STATE_6.0  _STATE_9.0  _STATE_12.0  _STATE_13.0  \
2087             1           0           0            0            0   
2088             1           0           0            0            0   
2090             1           0           0            0            0   
2091             1           0           0            0            0   
2093             1           0           0            0            0   
...            ...         ...         ...          ...          ...   
431675           0           0           0            0            0   
431676           0           0           0            0            0   
431677           0           0           0            0            0   
431680           0           0           0            0            0   
431682           0           0           0            0            0   

        _STATE_15.0  _STATE_17.0  _STATE_18.0  _STATE_20.0  _STATE_21.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_23.0  _STATE_24.0  _STATE_25.0  _STATE_26.0  _STATE_27.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_28.0  _STATE_29.0  _STATE_30.0  _STATE_31.0  _STATE_33.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_34.0  _STATE_35.0  _STATE_36.0  _STATE_39.0  _STATE_42.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0       

In [36]:
duplicate_columns = df_categorical.columns[df_categorical.columns.duplicated()]
duplicate_columns

Index([], dtype='object')

In [37]:
df_categorical.to_parquet("categoricalWithDummies.parquet")

The shape is (355111, 495) of the categorical with dummies.

In [38]:
df_numerical = (df_numerical - df_numerical.mean()) / df_numerical.std()

In [39]:
df_numerical.head()

_LLCPWT  _DRNKWK2   HHADULT  MENTHLTH     _BMI5      HTM4  _WT2RAKE  \
2087  0.298793 -0.137820 -0.225114 -1.409517  1.472739 -0.749125  2.794101   
2088  1.394371 -0.147597 -0.093554 -0.867187 -0.004431  0.202746  2.794101   
2090  1.270236 -0.147597 -0.093554  0.705573  0.629541 -1.320247  2.794101   
2091  1.103853 -0.143670 -0.093554  0.705573 -0.065783 -0.082815  2.794101   
2093  1.025322  0.261890  0.169566  0.705573 -1.297545  0.393120  2.794101   

        _STSTR   ALCDAY4    _STRWT  DROCDY4_  SLEPTIM1     HTIN4  _LLCPWT2  \
2087 -2.079255 -0.823026  2.794101 -0.117906  0.053732 -0.791858  1.120740   
2088 -2.079255  1.091122  2.794101 -0.301975 -0.241227  0.177143  1.252981   
2090 -2.079255  1.091122  2.794101 -0.301975 -0.241227 -1.276358  1.120740   
2091 -2.079255 -0.831433  2.794101 -0.226182 -0.093747 -0.065107  1.120740   
2093 -2.079248 -0.752962  2.794101  0.780781 -0.241227  0.419393  1.120740   

       HEIGHT3  PHYSHLTH     WTKG3   WEIGHT2      _PSU    _AGE80  
2087 -0.442285 -1.498830  0.780356  0.076879 -1.142471 -1.016613  
2088 -0.338735 -0.776594  0.069398 -0.025487 -1.142257  0.380796  
2090 -0.494060  0.774874 -0.145741 -0.056507 -1.141831 -0.118279  
2091 -0.364623 -1.525579 -0.102998 -0.050303 -1.141618 -1.415873  
2093 -0.312848 -1.552329 -1.007724 -0.180588 -1.141191 -0.717168

In [40]:
df_numerical.to_parquet("numericalStandardized2022.parquet")

In [41]:
test_num = pd.read_parquet("numericalStandardized2022.parquet")
test_num

_LLCPWT  _DRNKWK2   HHADULT  MENTHLTH     _BMI5      HTM4  _WT2RAKE  \
2087    0.298793 -0.137820 -0.225114 -1.409517  1.472739 -0.749125  2.794101   
2088    1.394371 -0.147597 -0.093554 -0.867187 -0.004431  0.202746  2.794101   
2090    1.270236 -0.147597 -0.093554  0.705573  0.629541 -1.320247  2.794101   
2091    1.103853 -0.143670 -0.093554  0.705573 -0.065783 -0.082815  2.794101   
2093    1.025322  0.261890  0.169566  0.705573 -1.297545  0.393120  2.794101   
...          ...       ...       ...       ...       ...       ...       ...   
431675 -0.371768 -0.139240 -0.225114  0.705573  0.148162 -0.273190 -0.430260   
431676  0.037413 -0.118348 -0.093554  0.705573 -0.282876 -0.273190 -0.430260   
431677  0.082288 -0.122527  0.038006  0.705573  0.387278  0.869055 -0.430260   
431680  0.012818 -0.141748 -0.093554  0.705573 -0.808301 -1.320247 -0.226143   
431682 -0.324736 -0.137820 -0.093554  0.705573 -0.482663  0.202746 -0.450114   

          _STSTR   ALCDAY4    _STRWT  DROCDY4_  SLEPTIM1     HTIN4  _LLCPWT2  \
2087   -2.079255 -0.823026  2.794101 -0.117906  0.053732 -0.791858  1.120740   
2088   -2.079255  1.091122  2.794101 -0.301975 -0.241227  0.177143  1.252981   
2090   -2.079255  1.091122  2.794101 -0.301975 -0.241227 -1.276358  1.120740   
2091   -2.079255 -0.831433  2.794101 -0.226182 -0.093747 -0.065107  1.120740   
2093   -2.079248 -0.752962  2.794101  0.780781 -0.241227  0.419393  1.120740   
...          ...       ...       ...       ...       ...       ...       ...   
431675  1.644339 -1.114492 -0.430260 -0.150389 -0.093747 -0.307357 -0.126412   
431676  1.644346 -0.795000 -0.430260  0.239403  0.053732 -0.307357  0.128715   
431677  1.644339 -1.108886 -0.430260  0.163610  0.053732  0.903894 -0.126412   
431680  1.643795 -0.828631 -0.226143 -0.193699 -0.241227 -1.276358  0.111935   
431682  1.643926 -0.823026 -0.450114 -0.117906  0.053732  0.177143 -0.252250   

         HEIGHT3  PHYSHLTH     WTKG3   WEIGHT2      _PSU    _AGE80  
2087   -0.442285 -1.498830  0.780356  0.076879 -1.142471 -1.016613  
2088   -0.338735 -0.776594  0.069398 -0.025487 -1.142257  0.380796  
2090   -0.494060  0.774874 -0.145741 -0.056507 -1.141831 -0.118279  
2091   -0.364623 -1.525579 -0.102998 -0.050303 -1.141618 -1.415873  
2093   -0.312848 -1.552329 -1.007724 -0.180588 -1.141191 -0.717168  
...          ...       ...       ...       ...       ...       ...  
431675 -0.390510 -1.525579 -0.038409 -0.040997  0.814725  0.580426  
431676 -0.390510  0.774874 -0.404573 -0.093731  0.814939 -0.018464  
431677 -0.261073 -1.445331  0.823574  0.083083  0.815152 -1.216243  
431680 -0.494060 -1.552329 -1.222863 -0.211608  0.815791 -1.515688  
431682 -0.338735  0.774874 -0.361355 -0.087527  0.816218  0.879871  

[122820 rows x 20 columns]

In [42]:
test_cat = pd.read_parquet("categoricalWithDummies.parquet")
test_cat

_STATE_1.0  _STATE_6.0  _STATE_9.0  _STATE_12.0  _STATE_13.0  \
2087             1           0           0            0            0   
2088             1           0           0            0            0   
2090             1           0           0            0            0   
2091             1           0           0            0            0   
2093             1           0           0            0            0   
...            ...         ...         ...          ...          ...   
431675           0           0           0            0            0   
431676           0           0           0            0            0   
431677           0           0           0            0            0   
431680           0           0           0            0            0   
431682           0           0           0            0            0   

        _STATE_15.0  _STATE_17.0  _STATE_18.0  _STATE_20.0  _STATE_21.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_23.0  _STATE_24.0  _STATE_25.0  _STATE_26.0  _STATE_27.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_28.0  _STATE_29.0  _STATE_30.0  _STATE_31.0  _STATE_33.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_34.0  _STATE_35.0  _STATE_36.0  _STATE_39.0  _STATE_42.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0       

In [43]:
df_proc = pd.concat([df_categorical, df_numerical, target], axis=1)
df_proc

_STATE_1.0  _STATE_6.0  _STATE_9.0  _STATE_12.0  _STATE_13.0  \
2087             1           0           0            0            0   
2088             1           0           0            0            0   
2090             1           0           0            0            0   
2091             1           0           0            0            0   
2093             1           0           0            0            0   
...            ...         ...         ...          ...          ...   
431675           0           0           0            0            0   
431676           0           0           0            0            0   
431677           0           0           0            0            0   
431680           0           0           0            0            0   
431682           0           0           0            0            0   

        _STATE_15.0  _STATE_17.0  _STATE_18.0  _STATE_20.0  _STATE_21.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_23.0  _STATE_24.0  _STATE_25.0  _STATE_26.0  _STATE_27.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_28.0  _STATE_29.0  _STATE_30.0  _STATE_31.0  _STATE_33.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0            0   
...             ...          ...          ...          ...          ...   
431675            0            0            0            0            0   
431676            0            0            0            0            0   
431677            0            0            0            0            0   
431680            0            0            0            0            0   
431682            0            0            0            0            0   

        _STATE_34.0  _STATE_35.0  _STATE_36.0  _STATE_39.0  _STATE_42.0  \
2087              0            0            0            0            0   
2088              0            0            0            0            0   
2090              0            0            0            0            0   
2091              0            0            0            0            0   
2093              0            0            0            0       

In [44]:
duplicate_columns = df_proc.columns[df_proc.columns.duplicated()]
duplicate_columns

Index([], dtype='object')

In [45]:
df_proc.to_parquet("processed_Dummies_Stand_2022.parquet")